File: Swap.iOS/Pages/CBPeripheralViewController.cs

using System;
using System.Drawing;

using MonoTouch.Foundation;
using MonoTouch.UIKit;
using MonoTouch.CoreBluetooth;

namespace Swap.iOS
{
	public partial class CBPeripheralViewController : UIViewController
	{
		public SampleCBPeripheralManagerDelegate pDelegate;
		public static CBPeripheralManager peripheralManager;
		public static CBMutableCharacteristic transferCharacteristic;
		public static NSData dataToSend;
		public static int sendDataIndex;

		public CBPeripheralViewController() : base ( "CBPeripheralViewController", null )
		{
		}

		public override void DidReceiveMemoryWarning()
		{
			// Releases the view if it doesn't have a superview.
			base.DidReceiveMemoryWarning();
			
			// Release any cached data, images, etc that aren't in use.
		}

		public override void ViewDidLoad()
		{
			base.ViewDidLoad();
			
			// Perform any additional setup after loading the view, typically from a nib.
			pDelegate = new SampleCBPeripheralManagerDelegate();
			peripheralManager = new CBPeripheralManager( pDelegate, null );
		}

		public override void ViewWillDisappear()
		{
			peripheralManager.StopAdvertising();
			base.ViewWillDisappear();
		}

		public static void sendData ()
		{
			bool sendingEOM = false;

			// end of message?
			if (sendingEOM) {
				bool didSend = peripheralManager.UpdateValue ("EOM", transferCharacteristic, null);

				if (didSend) {
					// it did, so mark it as sent
					sendingEOM = false;
				}
				// didn't send, so we'll exit and wiat for peripheralManagerIsReadyToUpdateSubscribers to call sendData again
				return; 
			}

			// We're sending data
			// Is there any left to send?
			if (sendDataIndex >= dataToSend.Length) {
				// No data left. Do nothing
				return;
			}

			// There's data left, so send until the callback fails, or we're done
			bool isSent = true;

			while (isSent) {
				// Word out how big it should be
				int amountToSend = (int)dataToSend.Length - sendDataIndex;

				// Can't be longer than 20 bytes
				if (amountToSend > SERVICES.NOTIFY_MTU)
					amountToSend = SERVICES.NOTIFY_MTU; 

				// Copy out the data we want
				NSData chunk = new NSData (dataToSend.Bytes + sendDataIndex, amountToSend);

				isSent = peripheralManager.UpdateValue (chunk, transferCharacteristic, null);

				// If it didn't work, drop out and wait for the callback
				if (!isSent)
					return;

				NSString stringFromData = new NSString (chunk, NSStringEncoding.UTF8);
				Console.WriteLine ("Sent: {0}", stringFromData);

				// It did send, so update out index
				sendDataIndex += amountToSend;

				// Was it the last one?
				if (sendDataIndex >= dataToSend.Length) {
					// Set this so if the send fails, we'll send it next time
					sendingEOM = true;

					bool eomSent = peripheralManager.UpdateValue( "EOM", transferCharacteristic, null );

					if ( eomSent ) {
						// It sent, we're all done
						sendingEOM = false;
						Console.WriteLine ("Sent: EOM");
					}

					return;
				}
			}
		}

		public static void peripheralManagerIsReadyToUpdateSubscribers( CBPeripheralManager peripheral ) {
			sendData();
		}
	
		class SampleCBPeripheralManagerDelegate : CBPeripheralManagerDelegate
		{
			public void StateUpdated( CBPeripheralManager peripheral )
			{
				if ( peripheral.State != CBPeripheralManagerState.PoweredOn ) {
					return;
				}

				if ( peripheral.State == CBPeripheralManagerState.PoweredOn ) {
					transferCharacteristic = new CBMutableCharacteristic( CBUUID.FromString( SERVICES.TRANSFER_CHARACTERISTIC_UUID ), CBCharacteristicProperties.Notify, null, CBAttributePermissions.Readable );

					CBMutableService transferService = new CBMutableService( CBUUID.FromString( SERVICES.TRANSFER_SERVICE_UUID ), true );

					CBMutableCharacteristic[] tmpArr = { transferCharacteristic };
					transferService.Characteristics = tmpArr;
					peripheralManager.AddService( transferService );
				}

				NSMutableDictionary dict = new NSMutableDictionary();
				dict.Add( CBAdvertisement.DataServiceUUIDsKey, CBUUID.FromString( SERVICES.TRANSFER_SERVICE_UUID ) );
				peripheralManager.StartAdvertising( dict );
			}

			public void CharacteristicSubscribed( CBPeripheralManager peripheral, CBCentral central, CBCharacteristic characteristic )
			{
				// TODO: dataToSend may need to be encoded appropriately
				dataToSend = textView.Text;
				sendDataIndex = 0;
				sendData();
			}
		}
	}
}

File: CoreBluetooth-master/CBTutorial/CBPeripheralViewController.m
//
//  CBPeripheralViewController.m
//  CBTutorial
//
//  Created by Orlando Pereira on 10/16/13.
//  Copyright (c) 2013 Mobiletuts. All rights reserved.
//

#import "CBPeripheralViewController.h"


@implementation CBPeripheralViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    _peripheralManager = [[CBPeripheralManager alloc] initWithDelegate:self queue:nil];
    _sendingEOM = NO;
}

- (void)viewWillDisappear:(BOOL)animated {
    [_peripheralManager stopAdvertising];
    
    [super viewWillDisappear:animated];
}

- (IBAction)sendButtonPressed:(UIButton *)sender
{
    NSLog(@"Button Tapped!");
    
    _dataToSend = [_textView.text dataUsingEncoding:NSUTF8StringEncoding];
    
    _sendDataIndex = 0;
    
    [self sendData];
}

- (void)peripheralManagerDidUpdateState:(CBPeripheralManager *)peripheral {
    NSLog(@"PeripheralViewController: peripheralManagerDidUpdateState");
    if (peripheral.state != CBPeripheralManagerStatePoweredOn) {
        return;
    }
    
    if (peripheral.state == CBPeripheralManagerStatePoweredOn) {

        _transferCharacteristic = [[CBMutableCharacteristic alloc] initWithType:[CBUUID UUIDWithString:TRANSFER_CHARACTERISTIC_UUID] properties:CBCharacteristicPropertyNotify value:nil permissions:CBAttributePermissionsReadable];
        
        CBMutableService *transferService = [[CBMutableService alloc] initWithType:[CBUUID UUIDWithString:TRANSFER_SERVICE_UUID]primary:YES];
        
        transferService.characteristics = @[_transferCharacteristic];
        
        [_peripheralManager addService:transferService];
    }
    
    [_peripheralManager startAdvertising:@{ CBAdvertisementDataServiceUUIDsKey : @[[CBUUID UUIDWithString:TRANSFER_SERVICE_UUID]] }];
}



- (void)peripheralManager:(CBPeripheralManager *)peripheral central:(CBCentral *)central didSubscribeToCharacteristic:(CBCharacteristic *)characteristic {
    NSLog(@"PeripheralViewController: didSubscribeToCharacteristic");
//    _dataToSend = [_textView.text dataUsingEncoding:NSUTF8StringEncoding];
//
//    _sendDataIndex = 0;
//
//    [self sendData];
}

- (void)sendData {
    NSLog(@"PeripheralViewController: sendData");
//    static BOOL sendingEOM = NO;
    
    // end of message?
    if (_sendingEOM) {
        NSLog(@"sendingEOM");
        BOOL didSend = [self.peripheralManager updateValue:[@"EOM" dataUsingEncoding:NSUTF8StringEncoding] forCharacteristic:self.transferCharacteristic onSubscribedCentrals:nil];
        
        if (didSend) {
            NSLog(@"did send");
            // It did, so mark it as sent
            _sendingEOM = NO;
            
        }
        // didn't send, so we'll exit and wait for peripheralManagerIsReadyToUpdateSubscribers to call sendData again
        NSLog(@"sendingEOM: return");
        return;
    }
    
    // We're sending data
    // Is there any left to send?
    if (self.sendDataIndex >= self.dataToSend.length) {
        // No data left.  Do nothing
        NSLog(@"no data left");
        return;
    }
    
    // There's data left, so send until the callback fails, or we're done.
    BOOL didSend = YES;
    
    while (didSend) {
        NSLog(@"didSend");
        // Work out how big it should be
        NSInteger amountToSend = self.dataToSend.length - self.sendDataIndex;
        
        // Can't be longer than 20 bytes
        if (amountToSend > NOTIFY_MTU) amountToSend = NOTIFY_MTU;
        
        // Copy out the data we want
        NSData *chunk = [NSData dataWithBytes:self.dataToSend.bytes+self.sendDataIndex length:amountToSend];
        
        didSend = [self.peripheralManager updateValue:chunk forCharacteristic:self.transferCharacteristic onSubscribedCentrals:nil];
        
        // If it didn't work, drop out and wait for the callback
        if (!didSend) {
            NSLog(@"didn't send: return");
            return;
        }
        
        NSString *stringFromData = [[NSString alloc] initWithData:chunk encoding:NSUTF8StringEncoding];
        NSLog(@"Sent: %@", stringFromData);
        
        // It did send, so update our index
        self.sendDataIndex += amountToSend;
        
        // Was it the last one?
        if (self.sendDataIndex >= self.dataToSend.length) {
            NSLog(@"sendDataIndex >= dataToSend.length");
            // Set this so if the send fails, we'll send it next time
            _sendingEOM = YES;

            BOOL eomSent = [self.peripheralManager updateValue:[@"EOM" dataUsingEncoding:NSUTF8StringEncoding] forCharacteristic:self.transferCharacteristic onSubscribedCentrals:nil];
            
            if (eomSent) {
                // It sent, we're all done
                _sendingEOM = NO;
                NSLog(@"Sent: EOM");
            }
            
            return;
        }
    }
}


- (void)peripheralManagerIsReadyToUpdateSubscribers:(CBPeripheralManager *)peripheral {
    NSLog(@"PeripheralViewController: peripheralManagerIsReadyToUpdateSubscribers");
    [self sendData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

@end
